Fix and generalize packet terraform scripts

joestringer opened the pull request:
Give the packet.net terraform scripts some love:
* Fix the default instance type so it can actually run 2-VM CI runs
* Fix terraform syntax errors
* Support running outside the Cilium CI instances
* Support both the previous shared keys model for provisioning, or alternatively just using the developer's SSH agent to authenticate via new `TF_VAR_shared_key` boolean
* Fix broken provision script due to missing vagrant preload script (only available on CI instances)
* Format the scripts more consistently
* Update the readme

(Validated through my own packet.net project, haven't tried with shared keys for access into CI environment)

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8996)
<!-- Reviewable:end -->

joestringer committed:
packet: Fix packet terraform scripts

Terraform complains:

    $ terraform apply
    var.packet_project_id
      Enter a value: xxx

    var.packet_token
      Enter a value: yyy

    var.private_key_path
      Enter a value: zzz

    Error: packet_device.test: "facilities": required field is not set

    Error: packet_device.test: "facility": [REMOVED] Use the "facilities" array instead, i.e. change
      facility = "ewr1"
    to
      facilities = ["ewr1"]

Fix this failure.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
packet: Default to x1.small.x86

The previous setting was somehow defaulting to the atom processors,
t1.small.x86, which only had 8GB of RAM, causing problems with the
default VM memory configuration of the VMs which lead to random
OOM-kills of the VMs while attempting to run ginkgo. x1.small.x86 is
almost 5x more expensive, but you'll save that much time having an
instance that actually works.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
packet: Format packet terraform script consistently

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
packet: Support disabling shared key

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
packet: Flesh out instructions for private usage

Previous instructions depended on shared keys and didn't point out one
or two steps. Fix them up and format them using standard GitHub
markdown.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
packet: Fix host reference

Fixes the following error:

    Error: Missing required argument

      on main.tf line 44, in resource "packet_device" "test":
      44:     connection {

    The argument "host" is required, but no definition was found.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
packet: Format script with functions

This makes the subsections clearer and easier to reason about.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
packet: Remove dependency on vagrant box preload

Allow this step to fail, since it relies on a script which may or may
not be available depending on whether you run directly on a node
provisioned by the CI, or a general packet.net node.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on test/packet/README.md:
```diff
@@ -1,28 +1,49 @@
-# 14 - Once tests are running and if you are running `screen`, you can leave the terminal
-#      by typing `CTRL+a+d`, to resume again type `screen -r`
-#
+1. Download terraform https://www.terraform.io/downloads.html
```
"How to run the CI tests in your own packet.net instances in 1 single step" (Check the 1. in all lines)

joestringer replied:
Oh, on GitHub markdown it automatically enumerates. Kinda forgot that most editors won't render it that way :)

aanm commented on test/packet/scripts/install.sh:
```diff
@@ -10,68 +10,86 @@ VAGRANT_VERSION="2.2.4"
 PACKER_VERSION="1.3.5"
```
Don't forget to push this upstream ;-)

joestringer replied:
Is this upstream? https://github.com/cilium/packer-ci-build/blob/master/provision/ubuntu/install.sh

joestringer committed:
packet: Add step for preventing vagrant abort

Vagrant will abort a VM run if memory is exhausted, for instance if the
tiny VM type is used (8GB) and MEMORY is 4G (the default). Suggest
overriding with 3GB to mitigate this.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
packet: Fix numbering outside of github

The previous numbering would render correctly on github, but users
reading from a standard editor would see just the number one lots of
times. Fix it.

Signed-off-by: Joe Stringer <joe@cilium.io>
